core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/ShardingContainerPoolBalancer.scala

```diff
@@ -396,38 +430,54 @@ object ShardingContainerPoolBalancer extends LoadBalancerProvider {
    * @return an invoker to schedule to or None of no invoker is available
    */
   @tailrec
-  def schedule(
-    maxConcurrent: Int,
-    fqn: FullyQualifiedEntityName,
-    invokers: IndexedSeq[InvokerHealth],
-    dispatched: IndexedSeq[NestedSemaphore[FullyQualifiedEntityName]],
-    slots: Int,
-    index: Int,
-    step: Int,
-    stepsDone: Int = 0)(implicit logging: Logging, transId: TransactionId): Option[(InvokerInstanceId, Boolean)] = {
+  def schedule(maxConcurrent: Int,
+               fqn: FullyQualifiedEntityName,
+               invokers: IndexedSeq[InvokerHealth],
+               dispatched: IndexedSeq[NestedSemaphore[FullyQualifiedEntityName]],
+               slots: Int,
+               index: Int,
+               step: Int,
+               stepsDone: Int = 0,
+               preferredInvokers: Seq[Int] = Seq.empty[Int])(
+    implicit logging: Logging,
+    transId: TransactionId): Option[(InvokerInstanceId, Boolean)] = {
     val numInvokers = invokers.size
 
     if (numInvokers > 0) {
-      val invoker = invokers(index)
-      //test this invoker - if this action supports concurrency, use the scheduleConcurrent function
-      if (invoker.status.isUsable && dispatched(invoker.id.toInt).tryAcquireConcurrent(fqn, maxConcurrent, slots)) {
-        Some(invoker.id, false)
+      if (preferredInvokers.nonEmpty) {
+        val preferredInvoker = preferredInvokers
+          .find(
+            invokerIndex =>
+              invokerIndex < invokers.size && invokers(invokerIndex).status.isUsable && dispatched(
+                invokers(invokerIndex).id.toInt).tryAcquireConcurrent(fqn, maxConcurrent, slots))
+          .map(invokerIndex => Some((invokers(invokerIndex).id, false)))
+        if (preferredInvoker.nonEmpty) {
+          preferredInvoker.get
+        } else {
+          schedule(maxConcurrent, fqn, invokers, dispatched, slots, index, step, stepsDone + 1)
```

Comment:
I think this will mess up the current home invoker algorithm in the case of complete overload because not all invokers may be checked. We should probably not add to steps done.

Reply:
yea that's right I missed that. This should be identical to the initial `schedule` call.